### PR TITLE
Use isolates form API and change model types

### PR DESCRIPTION
**Pull request by MicheleK17**
Ticket: #199
Change-Id: I07494d560616718c95f5005be8ca7a30eb3eca2a

### CODE REVIEW

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx [hunk 1]
```diff
@@ -15,28 +19,19 @@ import { getFilterFromPath } from "../../../Core/getFilterFromPath.service";
...
-}
-
 export function QueryPageContainerComponent(): JSX.Element {
     const { data, setData } = useContext(DataContext);
     const { filter, setFilter } = useContext(FilterContext);
     const { table, setTable } = useContext(TableContext);
     const history = useHistory();
 
-    const BASE_URL: string = mockDataURL;
+    const BASE_URL: string = isolateURL;
```
Comment (GerberI) [inline src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:28]: ISOLATE_URL?

File: src/app/Shared/Model/Isolate.model.ts [hunk 2]
```diff
@@ -1,3 +1,175 @@
...
+                value: string;
+                active: boolean;
+            };
+        };
+    }[];
+}
 
-export type DBentry = Record<DBkey, string>;
+
+export type DBentry = IsolateApiInterface["isolates"];
```
Comment (GerberI) [inline src/app/Shared/Model/Isolate.model.ts]: Reject: your business logic model should not depend on the api model,  i think your actual app does not interpret all the stuff in the api model? So keep an independent model for your app logic and map the api model to that model (only the data that is necessary and to a format that fits your components inputs)

Comment (GerberI) [reply src/app/Shared/Model/Isolate.model.ts]: Reject2: your dbentry is no entry anymore, it is a collection of the dbentries, so you should rename it

File: src/app/Shared/Model/Isolate.model.ts [hunk 3]
```diff
@@ -1,3 +1,175 @@
...
+
+export type DBkey = keyof IsolateApiInterface["isolates"][0];
```
Comment (GerberI) [inline src/app/Shared/Model/Isolate.model.ts]: Reject: see dbentry

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx [hunk 4]
```diff
@@ -15,28 +19,19 @@ import { getFilterFromPath } from "../../../Core/getFilterFromPath.service";
...
     const FILTER_URL: string = filterURL;
 
     const getData = async (): Promise<void> => {
-        const r: Response = await fetch(BASE_URL);
-        const dataProp: DBentry[] = await r.json();
-        const keyValueProps = Object.keys(dataProp[0]) as DBkey[];
+        const isolateResponse: Response = await fetch(BASE_URL);
+        const isolateProp: IsolateApiInterface = await isolateResponse.json();
+        const keyValueProps = Object.keys(isolateProp.isolates[0]) as DBkey[];
```
Comment (GerberI) [inline src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:34]: As stated in a previous review, consider refactoring this. Your data should not depend on the first entry of an api response. The client should know what it expects from the api and the data format should be clear for both sides, client and server.

File: src/app/Core/ExportServices/generateCSVString.service.ts [hunk 5]
```diff
@@ -14,14 +14,19 @@ function stringModification(inputString: string): string {
  * @returns {string} - data in one csv string
  */
 export function generateCSVString<
-    T extends Record<string, string>,
+    T extends Record<string, unknown>,
```
Comment (GerberI) [inline src/app/Core/ExportServices/generateCSVString.service.ts]: why is this unknown? Your code should be aware of how the data looks like it processes

**Review by GerberI — CHANGES_REQUESTED** [review]
Commit: Use isolates form API and change model types
In general the data format of the API does not look very flat. I recommend to keep the API as simple, flat and anambiguously as possible.

File: src/app/Core/ExportServices/generateRAWString.service.ts [hunk 6]
```diff
@@ -1,18 +1,17 @@
...
- * @param {DBentry} rawData - dataset
+ * @param {string[]} rawKeys - keys of the dataset
+ * @param {Record<string, string>[]} rawData - dataset
  * @returns {string} - dataset as string
  */
 export function RAWDataStringGenerator(
-    rawKeys: DBkey[],
-    rawData: DBentry
+    rawKeys: string[],
+    rawData: Record<string, string>[]
```
Comment (GerberI) [inline src/app/Core/ExportServices/generateRAWString.service.ts]: nice. Actually the stringGenerator is independent of the model.

File: src/app/Core/ExportServices/generateCSVString.service.ts (hunk 5 above)
Comment (MicheleK17) [reply src/app/Core/ExportServices/generateCSVString.service.ts]: Already changed it in the commit: "Adaption to isolate API "

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx [hunk 7]
```diff
@@ -36,6 +36,13 @@ export function QueryPageContainerComponent(): JSX.Element {
         const filterResponse: Response = await fetch(FILTER_URL);
         const filterProp: FilterConfigApiInterface = await filterResponse.json();
 
+        filterProp.filters.forEach((element, index) => {
+            if (element.id === "sContext") {
+                filterProp.filters[index].id = "samplingContext";
+                filterProp.filters[index].name = "samplingContext";
+            }
+        });
```
Comment (GerberI) [inline src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:44]: Reject: It is in general a bad idea to manipulating the data structure you are iterating over. I would suggest to map the filterProps to the desired datastructure. So const newFilterProps =  filterProp.filters.map(...)...

File: src/app/Shared/Model/Isolate.model.ts [hunk 8]
```diff
@@ -165,7 +165,7 @@ export interface IsolateApiInterface {
                 value: string;
                 active: boolean;
             };
-        };
+        }; */
```
Comment (GerberI) [inline src/app/Shared/Model/Isolate.model.ts:168]: See comment in review of previous commit

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx (hunk 7 above)
Comment (GerberI) [reply src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:44]: See comment in previous commit: Here you already mapping the api data to a more appropriate structure fitting your components desire. So you should use separate models for api and client data.

**Review by GerberI — CHANGES_REQUESTED** [review]
Commit: Use isolates form API and change model types

File: src/app/Core/ExportServices/generateFilteredDataString.service.ts [hunk 9]
```diff
@@ -6,14 +6,14 @@ import { generateCSVString } from "./generateCSVString.service";
  * @param {Record<string, string>[]} rawData - dataset
  * @returns {string} - dataset as string
  */
-export function RAWDataStringGenerator(
+export function generateFilteredDataString(
```
Comment (GerberI) [inline src/app/Core/ExportServices/generateFilteredDataString.service.ts:9]: not so good: now you renamed the universally usable function so it suggest to be only work with filteredData, but the name is better than the previous one

**Review by GerberI — APPROVED** [review]
Renaming inconsistent components
Some renaming is always good :)

File: src/app/Core/ExportServices/generateCSVString.service.ts (hunk 5 above)
Comment (GerberI) [reply src/app/Core/ExportServices/generateCSVString.service.ts]: then your typechecking code below is also obsolete?

**Review by GerberI — APPROVED** [review]
Commit: CR filterconfig-API
thx

**Review by GerberI — APPROVED** [review]
Commit: Renaming
thx,
i predict that this was not the last renaming of the Db... types :)

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx (hunk 7 above)
Comment (MicheleK17) [reply src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:54]: done. 
- 1. Reject in "Only use desired values from isolate-API"
- 2.Comment in "Separate api_isolate_model from data_isolate_model"

File: src/app/Core/ExportServices/generateCSVString.service.ts (hunk 5 above)
Comment (MicheleK17) [reply src/app/Core/ExportServices/generateCSVString.service.ts]: done. 
- Remove TypeCheck in generateCSVString

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx (hunk 4 above)
Comment (MicheleK17) [reply src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx]: Done. 
- Only use desired values from isolate-API

File: src/app/Shared/Model/Isolate.model.ts (hunk 8 above)
Comment (MicheleK17) [reply src/app/Shared/Model/Isolate.model.ts]: done.

File: src/app/Shared/Model/Isolate.model.ts (hunk 2 above)
Comment (MicheleK17) [reply src/app/Shared/Model/Isolate.model.ts]: done. 
- Only use desired values from isolate-API

File: src/app/Shared/Model/Isolate.model.ts (hunk 3 above)
Comment (MicheleK17) [reply src/app/Shared/Model/Isolate.model.ts]: done.

    Only use desired values from isolate-API

File: src/app/Shared/Model/Isolate.model.ts [hunk 10]
```diff
@@ -165,11 +164,16 @@ export interface IsolateDTO {
                 value: string;
                 active: boolean;
             };
-        }; */
+        };
```
Comment (GerberI) [inline src/app/Shared/Model/Isolate.model.ts]: Reject? Are these API changes?
The DTO should exactly reflect the data you get from the API.

File: src/app/Shared/Model/Isolate.model.ts [hunk 11]
```diff
@@ -165,11 +164,16 @@ export interface IsolateDTO {
...
+    microorganism: string;
+    samplingContext: string;
+    matrix: string;
+}[];
 
-export type DbCollection = IsolateDTO["isolates"];
+export type DbKey = "microorganism" | "samplingContext" | "matrix";
 
-export type DbKey = keyof IsolateDTO["isolates"][0];
+export const DbValues: DbKey[] = ["microorganism", "samplingContext", "matrix"];
```
Comment (GerberI) [inline src/app/Shared/Model/Isolate.model.ts]: DbKeys would be a better name

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx [hunk 12]
```diff
@@ -31,23 +32,32 @@ export function QueryPageContainerComponent(): JSX.Element {
...
         const isolateProp: IsolateDTO = await isolateResponse.json();
-        const keyValueProps = Object.keys(isolateProp.isolates[0]) as DbKey[];
+
+        const adaptedDbIsolates: DbCollection = isolateProp.isolates.map(
+            ({ microorganism, samplingContext, matrix }) => ({
+                microorganism,
+                samplingContext,
+                matrix,
+            })
+        );
```
Comment (GerberI) [inline src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:42]: Nice. Now if the api keys changes you just need to change this map function and the rest of your code works as before
Same for the filter props below

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx [hunk 13]
```diff
@@ -31,23 +32,32 @@ export function QueryPageContainerComponent(): JSX.Element {
...
 
         const filterResponse: Response = await fetch(FILTER_URL);
         const filterProp: FilterConfigDTO = await filterResponse.json();
 
-        filterProp.filters.forEach((element, index) => {
-            if (element.id === "sContext") {
-                filterProp.filters[index].id = "samplingContext";
-                filterProp.filters[index].name = "samplingContext";
+        const adaptedFilterProp = filterProp.filters.map(filterObj => {
+            const tempFilterProp = { ...filterObj};
```
Comment (GerberI) [inline src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx]: Consider using a more safe way to map the data. So explicitly create the new objects:
tempFilterProp = {
id: filterObj.id
name: filterObj.name
...
}
Another way would be using a json validator/schema during the json parse step.
Reason: Actually you don't know what the fetched data looks like. By typing it to the DTO you only tell the compiler how the data should look like. At runtime invalid data will be also mapped to your adaptedFilterProps and your app might crash somewhere else in the code. By explicitly mapping the values here your app will throw an exception also at his place, so you now whats going wrong.

**Review by GerberI — CHANGES_REQUESTED** [review]
Commit: Only use desired values from isolate-API
Much better now

**Review by GerberI — APPROVED** [review]
Commit: Separate api_isolate_model from data_isolate_model
I would move the api specific code to a separate sub feature. So then you don't need to name your model files client_model. In general your actual business logic things should have no amendments to their names. So API_model or IsolateDTO is fine because they are not part of your business logic.

**Review by GerberI — APPROVED** [review]
Commit: Remove TypeCheck in generateCSVString

File: src/app/Shared/Model/Client_Isolate.model.ts [hunk 14]
```diff
@@ -6,4 +6,4 @@ export type DbCollection = {
 
 export type DbKey = "microorganism" | "samplingContext" | "matrix";
 
-export const DbValues: DbKey[] = ["microorganism", "samplingContext", "matrix"];
+export const DbKeyCollection: DbKey[] = ["microorganism", "samplingContext", "matrix"];
```
Comment (GerberI) [inline src/app/Shared/Model/Client_Isolate.model.ts:9]: 👍 you can ignore my previous DbKeys suggestion

**Review by GerberI — APPROVED** [review]
Commit: Remove emptyFilter
nice

**Review by GerberI — APPROVED** [review]
Commit: Remove unused isTable-prop in Option branch
I almost forgot about that (but i did not :))
thx

File: src/app/Shared/Model/Isolate.model.ts (hunk 10 above)
Comment (MicheleK17) [reply src/app/Shared/Model/Isolate.model.ts]: As it is now, it is the current api.

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx (hunk 13 above)
Comment (MicheleK17) [reply src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx]: done. 
- Change how to adapted FilterProp

File: src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx [hunk 15]
```diff
@@ -13,13 +13,27 @@ import {
...
     FilterType,
+    SingleFilterConfig,
 } from "../../../Shared/Model/Filter.model";
 import { FilterContext } from "../../../Shared/Context/FilterContext";
 import { TableContext } from "../../../Shared/Context/TableContext";
 import { getFilterFromPath } from "../../../Core/getFilterFromPath.service";
 import { generatePathString } from "../../../Core/generatePathString.service";
 import { getFeaturesFromPath } from "../../../Core/getTableFromPath.service";
 
+function setAdaptedFilterProp(
```
Comment (GerberI) [inline src/app/Main/Body/QueryPage/QueryPage-Container.component.tsx:24]: getAdaptedFilterProp, but you can leave as is because it is a private function

**Review by GerberI — APPROVED** [review]
Commit: Change how to adapted FilterProp
Good, you also avoided duplicated code 👍

**Pull request merged**